devito/finite_differences/differentiable.py

```diff
@@ -0,0 +1,189 @@
+import sympy
+import numpy as np
+
+from devito.symbolics.extended_sympy import ExprDiv
+from devito.tools import filter_ordered
+from devito.finite_differences.utils import to_expr
+
+__all__ = ['Differentiable']
+
+
+class Differentiable(sympy.Expr):
+    """
+    This class represents Devito differentiable objects such as
+    sum of functions, product of function or FD approximation and
+    provides FD shortcuts for such expressions
+    """
+    # Set the operator priority higher than Sympy (10.0) to force the overloaded
+    # operators to be used
+    _op_priority = 100.0
+
+    def __new__(cls, *args, **kwargs):
+        if cls == Differentiable:
+            assert len(args) == 1
+            expr = args[0]
```

Comment:
I have an idea, I think. You can check here whether:
```
if expr.is_Function
```
in which case, you just
```
    return expr
```
this should avoid the proliferation of ``if self.is_Function`` in the various methods

Reply:
To avoid duplicates of `add,mul,...` makes evrything lots easier and less messy to have `Function` being differentiable